### PR TITLE
feat: support `UDPRoute` when expression router enabled

### LINKED ISSUE #4539: Support `UDPRoute` with expression router
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem Statement

Split out of #4312
Kong 3.4 added support of expression router in L4 proxy. We could implement support of `UDPRoute` in gateway APIs. 

### Proposed Solution

- Translate `UDPRoute` into expression based Kong route when `ExpressionRoutes` feature gate is enabled, and Kong uses expression router 

### Additional information

_No response_

### Acceptance Criteria

- [ ] KIC could translate `UDPRoute` to expression based Kong routes against Kong 3.4+ running with expression router

### PR DESCRIPTION
**Pull request by randmonkey**
**What this PR does / why we need it**:

support `UDPRoute` with expression router when Kong version is >3.4, and add tests for it.

**Which issue this PR fixes**:

fixes #4539 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/dataplane/parser/translate_udproute.go [hunk 1]
```diff
@@ -40,6 +49,11 @@ func (p *Parser) ingressRulesFromUDPRoutes() ingressRules {
 		}
 	}
 
+	// translate generated Kong Route to expression based route.
```
Comment (programmer04) [inline internal/dataplane/parser/translate_udproute.go:52]:
```suggestion
	// Translate generated Kong Route to expression based route.
```

File: internal/dataplane/parser/translate_udproute.go [hunk 2]
```diff
@@ -91,3 +90,20 @@ func (p *Parser) ingressRulesFromUDPRoute(result *ingressRules, udproute *gatewa
 
 	return nil
 }
+
+// validateUDPRoute validates UDPRoute, and return a translation error if the spec is invalid.
+// validation for UDPRoutes will happen at a higher layer, but in spite of that we run
```
Comment (programmer04) [inline internal/dataplane/parser/translate_udproute.go:95]:
```suggestion
// Validation for UDPRoutes will happen at a higher layer, but in spite of that we run
```

File: internal/dataplane/parser/translate_failures_test.go [hunk 3]
```diff
@@ -14,11 +14,17 @@ import (
...
 	"github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane/failures"
 	"github.com/kong/kubernetes-ingress-controller/v2/internal/store"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/versions"
 	kongv1beta1 "github.com/kong/kubernetes-ingress-controller/v2/pkg/apis/configuration/v1beta1"
 )
 
 // This file contains unit test functions to test translation failures genreated by parser.
 
+func newResourceFailure(reason string, objects ...client.Object) failures.ResourceFailure {
+	failure, _ := failures.NewResourceFailure(reason, objects...)
```
Comment (programmer04) [inline internal/dataplane/parser/translate_failures_test.go:24]: How are we sure that the error won't be returned and we just miss it?

**Review by programmer04 — CHANGES_REQUESTED** [review]

File: internal/dataplane/parser/translate_failures_test.go (hunk 3 above)
Comment (randmonkey) [reply internal/dataplane/parser/translate_failures_test.go]: Used `require` to fail the test if error happens.

File: internal/dataplane/parser/translate_tcproute.go [hunk 4]
```diff
@@ -25,6 +26,13 @@ func (p *Parser) ingressRulesFromTCPRoutes() ingressRules {
 
 	var errs []error
 	for _, tcproute := range tcpRouteList {
+		// Disable the translation to expression routes and register translation errors
+		// when expression route is enabled and Kong version is less than 3.4.
+		if p.featureFlags.ExpressionRoutes && p.kongVersion.Compare(versions.ExpressionRouterL4Cutoff) < 0 {
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply internal/dataplane/parser/translate_tcproute.go]: done.

**Review by programmer04 — APPROVED** [review]

**Pull request merged**